File: js/cash_flow.js
let activitySlice = {
    'sow_date': new Date('2022-04-01'),
    'harvest_date': new Date('2022-09-01'),
    'campaign_end_date': new Date('2022-06-01'),
    'campaign_start_date': new Date('2022-01-01')
};

let assignedCostItems = {
    'item_cost': 20000,
    'date_used': new Date('2021-12-01'),
    'category': 'semilla'
};

let paymentPolicy = [[75, "$sow_date-15"], [25, "$sow_date+15"]];

function extractDynamicDate (dynamicDate) {

    let pattern = /\$(?<period>harvest_date|sow_date|campaign_start_date|campaign_end_date)(?<op>\-|\+)(?<days>\d+)/g;
    let regex = new RegExp(pattern)
    let myArray = regex.exec(dynamicDate);
    return myArray

};

function validatePaymentPolicy (paymentPolicy) {

    let isListOfLen2 = paymentPolicy.every(function (t) {
        return t.length === 2
    });

    let isFirstInt = paymentPolicy.every(function (t) {
        return typeof(t[0]) === 'number'
    });

    let isSecondIntOrString = paymentPolicy.every(function (t) {
        
        let boolean = typeof(t[1]) === 'number' || typeof(t[1]) === 'string'
        return boolean
    });

    let sum = 0
    paymentPolicy.forEach(t => {
        sum += t[0]
    });

    let sumToAHundred = sum === 100;

    if (isSecondIntOrString && isListOfLen2) {
        for (let i = 0; i < paymentPolicy.length; i++) {
            let _ = extractDynamicDate(paymentPolicy[i][1])
        }
    };

    if (!isListOfLen2) {
        return [isListOfLen2, 'isListOfLen2']
    };

    if (!isFirstInt) {
        return [isFirstInt, 'isFirstInt']
    };

    if (!isSecondIntOrString) {
        return [isSecondIntOrString, 'isSecondIntOrString']
    };

    if (!sumToAHundred) {
        return [sumToAHundred, 'sumToAHundred']
    };

    return [true, '']
};

function createCashFlow (activitySlice, assignedCostItems, paymentPolicy) {

    let cashFlow = [];
    let totalCost = assignedCostItems['item_cost'];
    
    
    
    for(let i = 0; i< paymentPolicy.length; i++){
      let fraction = paymentPolicy[i][0];
      let dynamicDate = paymentPolicy[i][1];
      let paymentCost = fraction / 100*totalCost;
      let paymentDate;
      
      if(typeof dynamicDate == 'number'){
        dateUsed = assignedCostItems['date_used'];
        
        paymentDate = new Date()
        paymentDate.setDate(dateUsed.getDate() + dynamicDate);
        
      }
      if(typeof dynamicDate == 'string'){
        dynamicDate = extractDynamicDate(dynamicDate);
        let op = dynamicDate[2];
        let days = dynamicDate[3];
        let period = dynamicDate[1];
        
        
        if(op == '+'){
          let dateUsed = activitySlice[period];
          paymentDate = new Date();
          paymentDate.setDate(dateUsed.getDate() + days);
        }
        if(op == '-'){
          let dateUsed = activitySlice[period];
          paymentDate = new Date();
          paymentDate.setDate(dateUsed.getDate() - days);

        }
      }
    
    cashFlow.push([paymentDate, paymentCost]);
    }
  return cashFlow;

};

File: cash_flow.py
import datetime
from typing import List, Tuple
import re


class DynamicDateError(Exception):
    pass


def extract_dynamic_date(dynamic_date: str):

    expression = '\$(harvest_date|sow_date|campaign_start_date|campaign_end_date)(-|\+)(\d+)'
    result = re.search(expression, dynamic_date)

    if result == None:
        raise DynamicDateError

    else:
        col, op, days = result.groups()
        return (col, op, int(days))


def validate_payment_policy(payment_policy):
    """Validates if payment policies provided are structured in the desired way."""

    is_list_of_len_2 = all([len(t) == 2 for t in payment_policy])
    is_first_int = all([type(t[0]) == int for t in payment_policy])
    is_second_int_or_string = all([((type(t[1]) == str) or (
        type(t[1] == int))) for t in payment_policy])
    amount_to_a_hundred = sum([t[0] for t in payment_policy]) == 100

    if is_second_int_or_string and is_list_of_len_2:
        for _, dynamic_date in payment_policy:
            if type(dynamic_date) == str:
                _ = extract_dynamic_date(dynamic_date=dynamic_date)

    if not is_list_of_len_2:
        return (is_list_of_len_2, 'is_list_of_len_2')
    if not is_first_int:
        return (is_first_int, 'is_first_int')
    if not is_second_int_or_string:
        return (is_second_int_or_string, 'is_second_int_or_string')
    if not amount_to_a_hundred:
        return (amount_to_a_hundred, 'amount_to_a_hundred')

    return (True, '')


def create_cash_flow(activity_slice, assigned_cost_item, payment_policy):

    cash_flow = []
    total_cost = assigned_cost_item['item_cost']

    for fraction, dynamic_date in payment_policy:

        payment_cost = fraction / 100 * total_cost

        if type(dynamic_date) == int:
            payment_date = assigned_cost_item['date_used'] + \
                datetime.timedelta(days=dynamic_date)

        elif type(dynamic_date) == str:
            col, op, days = extract_dynamic_date(dynamic_date=dynamic_date)

            if op == '+':
                payment_date = activity_slice[col] + \
                    datetime.timedelta(days=days)
            elif op == '-':
                payment_date = activity_slice[col] - \
                    datetime.timedelta(days=days)

        cash_flow.append([payment_date, payment_cost])

    return cash_flow
